User manual documentation [11214] (#34)
eProsima/Fast-DDS-monitor: docs/rst/user_manual/chart_panel.rst

.. include:: ../exports/alias.include
.. include:: ../exports/roles.include

.. _chart_panel:

###########
Chart Panel
###########

> Functionality related with :ref:`main_panel_layout`.

Author: :+1:

Data Types
==========
A Data or *Data Point* in the Monitor refers to a specific data type of the DDS network monitoring.
Each data point consists in a time and a value.
The time is the moment the data has been created, reported, received, etc. depending on the *DataKind*.

> The value refers to the value of this *DataKind* in the particular moment that represents the time of the data.

Author: Me neither

i.e. each data point of the *DataKind* ``DATA_COUNT`` has a time and an integer value, and it is stored in a
*DataWriter*.
This data point value refers to the number of ``Data packages`` this *DataWriter* sent since the last time this data
was reported to the time this data has.

Hence, the representation of all the *DataKinds* will be similar. |br|
The X axis is the time value of the data. |br|
The Y axis is the value that is store in the data.

> This could be integers, doubles or times, but for the same *DataKind* will always be the same value type.

Author: :+1:

*Participants*.
Hence, the data displayed will be all the data that all those *DataWriters* have stored, accumulated by the
cumulative function.

It is recommended to check some examples (:ref:`start_tutorial`) in order to better understand this functionality.

Target Entity Id
----------------

.. note::

    Not every *DataKind* has a target, so the dialog will only show this field when the *DataKind* selected
    requires it.

This is the entity Id of the entity that the data refers to. |br|
This field works similar to the :ref:`source_entity_id`.
Some *DataKind* has a target entity that the data refers to, and this target must be of an specific entity kind.
Choosing an entity of a different kind than the one this data requires, will be solved by using the same
mechanism explained in :ref:`source_entity_id`.
That is, searching for the correct entity kind by following the connections between entities.

i.e. To show the *DataKind* ``FASTDDS_LATENCY``, that is stored in a *DataWriter*, and targeted to a *DataReader*,
from an entity ``Host_1`` to an entity ``Host_2``, both of kind *Host*, the data displayed would be collected
like this:

> - Get all the *DataWriters* of ``Host_1`` by searching for its *Users*, their *Process*, their *Participants* and

Reviewer: ```suggestion
- Get all the *DataWriters* of ``Host_1`` by searching for its *Users*, their *Processes*, their *DomainParticipants* and
```

  their respective *DataWriters*.

> - Get all the *DataReaders* of ``Host_2`` by searching for its *Users*, their *Process*, their *Participants* and

Reviewer: ```suggestion
- Get all the *DataReaders* of ``Host_2`` by searching for its *Users*, their *Processes*, their *Participants* and
```

  their respective *DataReaders*.
- Get all the data from the *DataWriters* found that refer to the *DataReaders* found inside the interval set.

> - Accumulate the data by the *cumulative function*.

Reviewer: ```suggestion
- Accumulate the data using the *cumulative function*.
```

*Fast DDS* applications for the period of two hours in intervals of ten minutes.

Chart series
============
A *DataPoint* in a chart series in the *Fast DDS Monitor* refers to a specific data type of the DDS network monitoring.
Each *DataPoint* consists in a timestamp and a real value.
The *timestamp* is the moment the data has been created, reported, received, etc. depending on the *DataKind*.
The *value* refers to the real value of this *DataPoint* for this *DataKind* at the moment of the *timestamp* value.

For example, each *DataPoint* of the *DataKind* ``DATA_COUNT`` has a timestamp and an integer value.
This value is the number of ``Data packets`` a *DataWriter* sent since the last time that same data was reported.

.. todo::

    For a complete description of each *DataKind*, please refer to ...
    Add link to backend documentation once there is a first version.

Chart view
==========
Every *DataKind* is represented similarly inside a *Chartbox*.
None, one or multiple lines of different colors will represent the different data series that are being displayed

> in the chart with the following format: |br|

Reviewer: ```suggestion
in the chart with the following format:
```

The X axis is the time value of the data. |br|

> The Y axis is the value that is store in the data.

Reviewer: This could be a list

> This could be integers, doubles or times, but for the same *DataKind* will always be the same value type. |br|

Reviewer: ```suggestion
This could be integers, doubles or times, but for the same *DataKind* will always be the same value type.
```

Every point represents a *DataPoint* of data in a particular time.

.. _create_serie:

Create Series Dialog
====================
This Dialog (:ref:`create_new_series_layout`) allows to create a new data series in a Chartbox.
The fields in the dialog configure the data that will be displayed

Series label
------------
Name the new series in the Chartbox.

> If not set, the default name of a series is
> ``<cumulative_function>_<source_entity_kind>_<target_entity_kind>-<target_entity_id>``.

Reviewer: ```suggestion
If not set, the default series name is formed according to the following rule:

``<cumulative_function>_<source_entity_kind>_<target_entity_kind>-<target_entity_id>``
```

- Get all the *DataWriters* of ``Host_1`` by searching for its *Users*, their *Processes*, their
  *DomainParticipants* and their respective *DataWriters*.
- Get all the *DataReaders* of ``Host_2`` by searching for its *Users*, their *Processes*, their
  *DomainParticipants* and their respective *DataReaders*.
- Get all the data from the *DataWriters* found that refer to the *DataReaders* found inside the interval set.
- Accumulate the data using the *cumulative function*.

It is recommended to check some examples (:ref:`start_tutorial`) in order to better understand this functionality.

.. todo::

    Add link to backend documentation once there is a first version.

.. _number_of_bins:

Number of bins
--------------
Number of *DataPoints* that will be displayed for this chart series.
The data is collected in individual points in every entity, without a regular time interval or pattern.
Therefore, to show the data in a better understandable manner, each of these points will be merged in a single point
inside a fraction of the time interval.
The *Number of bins* determines how many fractions on which this time interval will be split, and thus, how many points

> will be displayed in the chart. |br|

Reviewer: ```suggestion
will be displayed in the chart.
```

Reset zoom
^^^^^^^^^^
Reset the zoom of the Chartbox to the standard one.
The standard zoom is calculated to fit all the data that is being currently displayed.

Clear chart
^^^^^^^^^^^
Eliminate every data configuration displayed in the Chartbox

Remove chart
^^^^^^^^^^^^
Eliminate the Chartbox and every configuration in it.

Series Menu
-----------
In the top bar of each Chartbox there is a Menu tab *Series* with these buttons.

Add series
^^^^^^^^^^
Add a new series in this Chartbox.
This will open a new :ref:`create_serie`.

Hide all series
^^^^^^^^^^^^^^^
Hide all series in this Chartbox.

Display all series
^^^^^^^^^^^^^^^^^^
Reveal all series in this Chartbox.

Chart Interaction
-----------------
The user could interact with the Chartbox and the data in it by resizing and moving the view.

Point description
^^^^^^^^^^^^^^^^^
Clicking in any point of a displayed series, an info box will be prompt with the ``x`` and ``y``
label values of the data point.

Moving the view
^^^^^^^^^^^^^^^
Press and hold ``Ctrl`` key and left click in one point of the Chartbox.
Move the mouse to scroll over the data.

Zoom in/out
^^^^^^^^^^^

> Press and hold ``Ctrl`` key and scroll up to zoom in to the center of the Chartbox. |br|

Reviewer: ```suggestion
Press and hold ``Ctrl`` key and scroll up to zoom in to the center of the Chartbox.
```
